ext/atomic/index.md

```diff
@@ -0,0 +1,510 @@
+described below.
+
+### <a href="#result-objects" id="result-objects" class="headerlink"></a> Result Objects
+
+An operation result object **MAY** contain any of the following members:
+
+- `data`: the "primary data" resulting from the operation.
+
+- `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the result.
+
+An empty result object (`{}`) is acceptable for operations that are not required
+to return `data`.
+
+## <a href="#processing" id="processing" class="headerlink"></a> Processing
+
+All HTTP requests sent with this extension **MUST** be issued with `POST`.
+
+A server **MUST** perform operations in the order they appear in the
+`atomic:operations` array.
+
+A server **MUST** perform all operations atomically, so that a failure
+to perform any operation **MUST** invalidate any effects of preceding
+operations.
+
+A server **MUST** respond to a successful operations request with `200 OK` if a
+response document is returned. An array of [result objects](#result-objects)
+**MUST** be returned in a top-level `atomic:results` member. The results array
+**MUST** be the same length as the requested operations array and each result
+**MUST** correspond positionally to its associated operation.
+
+A server **MAY** respond to successful requests that include operations with
+`204 No Content` and no response document if no operations are required to
+return `data`.
+
+### <a href="#processing-errors" id="processing-errors" class="headerlink"></a> Processing Errors
+
+If any operation in a request fails, the server **MUST** respond as described
+[in the base specification](/format/#errors-processing). An array of one or
+more [error objects](/format/#error-objects) **SHOULD** be returned, each with
+a `source` member that contains a `pointer` to the source of the problem in the
+request document.
+
+### <a href="#specific-operations" id="specific-operations" class="headerlink"></a> Processing Specific Operations
+
+The following sections describe how to process specific operations.
+
+#### Creating Resources
+
+An operation that creates a resource **MAY** target a resource collection
+through the operation's `ref` member. The operation **MUST** include an `op`
+code of `"add"` as well as a resource object as `data`. The resource object
+**MUST** contain at least a `type` member.
+
+For example:
+
+```json
+POST /bulk HTTP/1.1
+Host: example.org
+Content-Type: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+Accept: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+
+{
+  "atomic:operations": [{
+    "op": "add",
+    "href": "/blogPosts",
+    "data": {
+      "type": "articles",
+      "attributes": {
+        "title": "JSON API paints my bikeshed!"
+      }
+    }
+  }]
+}
+```
+
+Note that `href` is used in this example to target a resource collection,
+`blogPosts` in this case, that is distinct from the `type` of the resource.
```

Comment:
is `href` optional?

Reply:
Maybe I misunderstand something in the base spec (I'm studying JSON API only shortly) but what does it mean if `href` isn't specified? Is the resource somehow created in the collection of the same name as `data.type`? What if that doesn't exist?

Reply:
I forgot to explain the reference to base spec: Is there generally any constraint (or guidance) about the URI of the collection API (e.g. `/articles`) and type(s) of resource objects it can contain?

Reply:
There are no statements about URL structure in the spec, there are some recommendations outside the spec though: https://jsonapi.org/recommendations/#urls-resource-collections

I like to use "reminders" as an example. I.e. a server could have a `/reminders/{id}` for GETing all reminders globally and _also_ `/user/{userID}/reminders` for GETing and POSTing reminders for only a specific user. All of that would be spec-compliant.

Reply:
What is the default value of `href` then? According to https://jsonapi.org/recommendations/#urls-resource-collections having resource collection with the same name as each type is only recommended (even though it makes great sense). So default value can't just be `/articles`, right?

On a related note, using your `reminders` example, is this scenario valid according to spec/recommendations?

- POSTing a reminder to `/user/{userID}/reminders` makes that reminder available also directly under `/reminders` (with the same `id`).

In different words, JSON API allows to model the concept of "views" (`/user/{userID}/reminders` is a view to `/reminders`).

Reply:
> What is the default value of href then?

Reply:
> On a related note, using your `reminders` example, is this scenario valid according to spec/recommendations? ...

Yes. That's valid.

Reply:
> > What is the default value of href then?
> 
> I don't think there needs to be a default. I think this should be part of the clarifications we make about `ref` & `href`

Earlier you said that the `href` is optional. If it's missing in which collection the article gets created?

Reply:
I'm sorry. I see how what I said could be confusing. `ref` and `href` are mutually exclusive and it's usually optional which one you use. If you use `ref`, then you can't use `href`. When using `ref` a newly created resource object will go into the collection for that resource object's type. Finally, in certain cases, you can omit _both_ `href` and `ref`. For example, let's say that you have a `/reminders/operations` endpoint. At that endpoint, all operations must modify or create only `reminder` resource objects. The sensible default for that endpoint would be to add new resource objects to the `/reminders` collection.

This extension (and the spec for that matter) tries not to be overly prescriptive. Your server should choose to do what makes sense. The philosophy of JSON:API is generally, "_if_ you implement this like that, then do it like this." I.e., _if_ you support operations on a generic endpoint like `/operations` then you _should_ use `ref` or `href` as necessary. However, if you do not need `ref` or `href` because your operations endpoint is limited in scope, like `/reminders/operations` then you are not required to use `href` or `ref`. That is why the language in the extension is:

> An operation object **MAY** contain [either `ref` or `href`]

instead of:

> An operation object **MUST** contain [either `ref` or `href`]

Reply:
Ok. I now understand. Thanks for your explanation! Obviously it would be good if this explanation appeared (in the distilled form) in the final documentation.